src/object/schedcontext.c

```diff
@@ -159,20 +159,32 @@ static inline void maybeStallSC(sched_context_t *sc)
 }
 #endif
 
-static inline void setConsumed(sched_context_t *sc, word_t *buffer)
+static inline void replyFromKernel_consumed(tcb_t *thread, time_t consumed)
+{
+    word_t *buffer = lookupIPCBuffer(true, thread);
+    setRegister(thread, badgeRegister, 0);
+    word_t length = mode_setTimeArg(0, consumed, buffer, thread);
+    setRegister(thread, msgInfoRegister, wordFromMessageInfo(seL4_MessageInfo_new(0, 0, 0, length)));
+}
+
+static inline void setConsumed(sched_context_t *sc, tcb_t *thread, bool_t write_msg)
 {
     time_t consumed = schedContext_updateConsumed(sc);
-    word_t length = mode_setTimeArg(0, consumed, buffer, NODE_STATE(ksCurThread));
-    setRegister(NODE_STATE(ksCurThread), msgInfoRegister, wordFromMessageInfo(seL4_MessageInfo_new(0, 0, 0, length)));
+
+    if (write_msg) {
+        replyFromKernel_consumed(thread, consumed);
+    }
 }
 
-static exception_t invokeSchedContext_Consumed(sched_context_t *sc, word_t *buffer)
+static exception_t invokeSchedContext_Consumed(sched_context_t *sc, bool_t call)
 {
-    setConsumed(sc, buffer);
+    setConsumed(sc, NODE_STATE(ksCurThread), call);
+
+    setThreadState(NODE_STATE(ksCurThread), ThreadState_Running);
```

Comment:
I'm a bit confused about when `setThreadState(NODE_STATE(ksCurThread), ThreadState_Running);` needs to be called and where. `handleInvocation` seems to change restart to running in the normal case, but it's behind an `unlikely`.